central/networkgraph/flow/datastore/internal/store/postgres/store.go

```diff
@@ -613,6 +665,44 @@ func (s *flowStoreImpl) RemoveOrphanedFlows(ctx context.Context, orphanWindow *t
 	return s.pruneFlows(ctx, pruneStmt, orphanWindow)
 }
 
+func (s *flowStoreImpl) pruneOrphanExternalEntities(ctx context.Context, srcFlows []*storage.NetworkFlow, dstFlows []*storage.NetworkFlow) error {
+	// srcFlows contains flows where src is the deployment,
+	// so prune external flows based on the dst entity
+	if len(srcFlows) != 0 {
+		err := utils.BatchProcess(srcFlows, orphanedEntitiesPruningBatchSize, func(flows []*storage.NetworkFlow) error {
+			entities := make([]string, 0, len(flows))
+			for _, flow := range flows {
+				entities = append(entities, flow.GetProps().GetDstEntity().GetId())
+			}
+
+			pruneStmt := fmt.Sprintf(pruneOrphanExternalNetworkEntitiesSrcStmt, s.partitionName)
+			return s.pruneEntities(ctx, pruneStmt, entities)
+		})
+		if err != nil {
+			return err
+		}
+	}
+
+	// dstFlows contains flows where dst is the deployment,
+	// so prune external flows based on the src entity
+	if len(dstFlows) != 0 {
+		err := utils.BatchProcess(dstFlows, orphanedEntitiesPruningBatchSize, func(flows []*storage.NetworkFlow) error {
+			entities := make([]string, 0, len(flows))
+			for _, flow := range flows {
+				entities = append(entities, flow.GetProps().GetSrcEntity().GetId())
+			}
+
+			pruneStmt := fmt.Sprintf(pruneOrphanExternalNetworkEntitiesDstStmt, s.partitionName)
+			return s.pruneEntities(ctx, pruneStmt, entities)
+		})
+		if err != nil {
+			return err
+		}
+	}
```

Comment:
There is also code repetition here that can be eliminated. Make this into a separate function where you can pass in the flows, either `dstFlows` or `srcFlows`, and pass in the query, either `pruneOrphanExternalNetworkEntitiesDstStmt` or `pruneOrphanExternalNetworkEntitiesSrcStmt`.

Reply:
There is a subtle difference in the attribute of the flow we copy, `.GetProps().GetSrcEntity().GetId()` vs `.GetProps().GetDstEntity().GetId()`. It would then require to pass a lambda as parameter as well.

I think that it would make the code harder to understand.

Reply:
You could pass in the entity ids rather than passing in the flows.

Reply:
That is true. I did not implement this because I wanted to avoid making a full copy of the entities ID list.

As discussed, there would be another interest with combining in an entities ID list: we could deduplicate those.

Reply:
I also avoid getting the flows and just get the ids.